src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,342 @@
+
+## Ready for part 2?
+
+A fun way to get to know the spec is to start with a JavaScript feature we know is there, and find out how it's specified.
+
+We know that properties are looked up in the prototype chain: if an object doesn't have the property we're trying to read, we walk up the prototype chain until we find it (or find an object which no longer has a prototype).
+
+For example:
+
+```javascript
+const o1 = {'foo' : 2511};
+const o2 = {};
+console.log(o2.foo); // undefined
+Object.setPrototypeOf(o2, o1);
+console.log(o2.foo); // 2511
+```
+
+## Where's the prototype walk defined?
+
+Let's try to find out where this behavior is defined. A good place to start is a list of [Object Internal Methods](https://tc39.es/ecma262/#sec-object-internal-methods-and-internal-slots).
+
+There's both `[[GetOwnProperty]]` and `[[Get]]` &mdash; now we're interested in the version that isn't restricted to _own_ properties, so we'll go with `[[Get]]`.
+
+Unfortunately, the [Property Descriptor specification type](https://tc39.es/ecma262/#sec-property-descriptor-specification-type) also has a field called `[[Get]]`, so while browsing the spec for `[[Get]]`, we need to carefully distinguish between the two independent usages.
+
+The internal method `[[Get]]` delegates to `OrdinaryGet`:
+
+> [`[[Get]] ( P, Receiver )`](https://tc39.es/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-get-p-receiver)
+>
+> When the `[[Get]]` internal method of `O` is called with property key `P` and ECMAScript language value `Receiver`, the following steps are taken:
+>
+> 1. Return `? OrdinaryGet(O, P, Receiver)`.
+
+We'll see shortly that `Receiver` is the value which is used as the **this value** when calling a getter function of an accessor property.
+
+`OrdinaryGet` is defined like this:
+
+> [`OrdinaryGet ( O, P, Receiver )`](https://tc39.es/ecma262/#sec-ordinaryget)
+>
+> When the abstract operation `OrdinaryGet` is called with Object `O`, property key `P`, and ECMAScript language value `Receiver`, the following steps are taken:
+>
+> 1. Assert: `IsPropertyKey(P)` is `true`.
+> 2. Let `desc` be `? O.[[GetOwnProperty]](P)`.
+> 3. If `desc` is `undefined`, then
+> a. Let `parent` be `? O.[[GetPrototypeOf]]()`.
+> b. If `parent` is `null`, return `undefined`.
+> c. Return `? parent.[[Get]](P, Receiver)`.
+> 4. If `IsDataDescriptor(desc)` is `true`, return `desc.[[Value]]`.
+> 5. Assert: `IsAccessorDescriptor(desc)` is `true`.
+> 6. Let `getter` be `desc.[[Get]]`.
+> 7. If `getter` is `undefined`, return `undefined`.
+> 8. Return `? Call(getter, Receiver)`.
+
+The prototype chain walk is inside step 3: if we don't find the property as an own property, we recurse into the prototype's `[[Get]]` method.
```

Comment:
Reworded this. As an non-native English speaker, it would never occur to me to use the word "recur" here, and I think it would be very surprising for other non-native speakers. Trying to minimize linguistic surprises here, so rewording it like this:

The prototype chain walk is inside step 3: if we don't find the property as an own property, we call the prototype's `[[Get]]` method. If we don't find the property there either, we call it's prototype's `[[Get]]` method, and so on, until we reach an object without a prototype.